src/microbe_stage/editor/EditorActionHistory.cs

```diff
@@ -138,8 +138,16 @@ public override bool Undo()
 
     public override void AddAction(TAction action)
     {
-        // TODO: check if the action can be combined (for example behaviour or rigidity slider subsequent edits should
+        // Check if the action can be combined (for example behaviour or rigidity slider subsequent edits should
         // combine) in a single step for undo
+        if (ActionIndex > 0 && action.Data.FirstOrDefault()!.WantsMerge)
+        {
+            EditorCombinableActionData currentActionData = action.Data.FirstOrDefault()!;
+            EditorCombinableActionData lastActionData = Actions[ActionIndex - 1].Data.FirstOrDefault()!;
```

Comment:
```suggestion
            EditorCombinableActionData currentActionData = action.Data.First();
            EditorCombinableActionData lastActionData = Actions[ActionIndex - 1].Data.First();
```

First returns a non-nullable value (or throws) so it is much better to be explicit about that than suppress nullability of the version that can return null (which would then hit a much less helpful null pointer exception).